syntax = "proto3";

package gammatrade;

// Action enum for trade signals
enum TradeAction {
    UNKNOWN = 0;
    BUY = 1;
    SELL = 2;
    HOLD = 3;
}

// OrderType enum
enum OrderType {
    MARKET = 0;
    LIMIT = 1;
    STOP = 2;
    STOP_LIMIT = 3;
}

// TimeInForce enum
enum TimeInForce {
    DAY = 0;
    GTC = 1;      // Good Till Cancelled
    IOC = 2;      // Immediate Or Cancel
    FOK = 3;      // Fill Or Kill
}

// TradeSignal message represents a trading signal generated by the Strategy Engine.
message TradeSignal {
    // Unique signal identifier
    string signal_id = 1;
    
    // Trading symbol
    string symbol = 2;
    
    // Action to take
    TradeAction action = 3;
    
    // Quantity to trade
    double quantity = 4;
    
    // Order type
    OrderType order_type = 5;
    
    // Limit price (for LIMIT orders)
    double limit_price = 6;
    
    // Stop price (for STOP orders)
    double stop_price = 7;
    
    // Time in force
    TimeInForce time_in_force = 8;
    
    // Target execution adapter (e.g., "alpaca", "coinbase")
    string target_adapter = 9;
    
    // Strategy that generated this signal
    string strategy_id = 10;
    
    // Confidence score (0.0 to 1.0)
    double confidence = 11;
    
    // Unix timestamp in milliseconds
    int64 timestamp = 12;
    
    // Optional: Reason/explanation for the signal
    string reason = 13;
    
    // Optional: Additional metadata
    map<string, string> metadata = 14;
}
